### ISSUE TITLE
Pausing/resuming worker causes process to crash

### ISSUE THREAD

**Issue by wenq1** [taskforcesh/bullmq#155]
Reproduction sequence:

```ts
var queueMqOptions     = {
        connection:             {host: 127.0.0.1, port: 6379},
};
var workermq = new Worker ('queue', funcA, queueMqOptions});
await workermq.waitUntilReady ();
await workermq.pause ();
await workermq.resume ();
```
Hangs after calling to workermq.resume, and after a few while of doing nothing, the process crashes with out-of-memory error

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x10072ee79]
Security context: 0x38769d5408a1 <JSObject>
    1: next [0x38769d563759](this=0x387653743039 <Array Iterator map = 0x38763e105ed1>)
    2: race [0x38769d547b11](this=0x38769d547821 <JSFunction Promise (sfi = 0x3876a6e4bec1)>,0x387653742e31 <JSArray[1]>)
    3: run [0x38764d497961] [~/test-server/node_modules/bullmq/dist/classes/worker.js:~60] [pc=0xcbf523736ab](this=0x3876dd9112c.

**Comment by wenq1** [comment]
The process is crashed due to infinite loop in worker.run ()

**Comment by manast** [comment]
I added this test case and it is working fine:
```ts
  it('should pause an empty queue locally', async () => {
    let worker: Worker;
    const process = async (job: Job) => {};

    worker = new Worker(queueName, process);
    await worker.waitUntilReady();
    await worker.pause();

    expect(worker.isPaused()).to.be.eql(true);

    worker.resume();

    await worker.close();
  });
```

**Comment by Wolvan** [comment]
I can definitely reproduce it. Attaching a debugger to my Node process and checking the heap after resuming I get this:
![image](https://user-images.githubusercontent.com/1437010/112989438-37462380-9165-11eb-9719-209cb750235f.png)
which honestly doesn't seem like a healthy heap, if you're asking me.
The heap seems to fill up with an Array of Arrays of Promises.

[Heap-20210330T150119.heapsnapshot.zip](https://github.com/taskforcesh/bullmq/files/6229199/Heap-20210330T150119.heapsnapshot.zip)
Attached is a snapshot of my heap before it ended up 4GB, but the culprit array is the biggest piece of the cake already.

**Comment by manast** [comment]
@Wolvan if you can reproduce it can you provide a minimal use case that reproduces it so that we can look into it?

**Comment by Wolvan** [comment]
Alright, here you go. Also recorded a video so you know exactly what I did.

```js
"use strict";

const Worker = require("bullmq").Worker;
const worker = new Worker("DEMO", async () => {
    console.log(`Doing some WORK, yo!`);
});

process.on("SIGUSR2", async () => {
    if (worker.isPaused()) {
        worker.resume();
        console.log("Worker resumed");
    } else {
        console.log("Pausing Worker...");
        await worker.pause();
        console.log("Worker paused");
    }
});
```
Versions: bullmq@1.10.0
MacOS: 10.14.6
NodeJS: 14.15.1
https://user-images.githubusercontent.com/1437010/113395682-845f0b00-939a-11eb-810e-d2bf6d7a79a8.mp4

Turns out I was on an outdated bullmq version. What I notice now instead is that trying to resume gives me a redis connection closed error, see
![image](https://user-images.githubusercontent.com/1437010/113396231-7a89d780-939b-11eb-9a02-e6603a830716.png)

So on 1.15.1 the memory leak is fixed, but resuming still seems to fail.

**Comment by wenq1** [comment]
Not sure if #389 and #359 is related to this.

The infinite loop is the root cause of the crash, not the resume.

**Comment by manast** [comment]
@Wolvan I am trying to write a test to verify that the fix indeed works, however while I can reproduce the issue using your code and instructions, if I write it like this it is not reproduced, any ideas?
```
'use strict';

const { Worker } = require('../dist');
const worker = new Worker('DEMO', async () => {
  console.log(`Doing some WORK, yo!`);
});

async function test() {
  process.on('SIGUSR2', async () => {
    if (worker.isPaused()) {
      worker.resume();
      console.log('Worker resumed');
    } else {
      console.log('Pausing Worker...');
      await worker.pause();
      console.log('Worker paused');

      setTimeout(() => process.kill(process.pid, 'SIGUSR2'), 3000);
    }
  });

  await worker.waitUntilReady();
  process.kill(process.pid, 'SIGUSR2');
}

test();
```

**Comment by wenq1** [comment]
issue is indeed in worker.whenCurrentJobsFinished ()

The following line should check if the connection is indeed disconnected before trying to reconnect. Don't reconnect based on the values of "this.waiting"!

this.waiting && reconnect && (await this.blockingConnection.reconnect());

**Comment by manast** [comment]
@wenq1 sure, this PR should fix it: https://github.com/taskforcesh/bullmq/pull/548 but we are missing a test to verify it...

**Comment by wenq1** [comment]
Closed as per #548.

Finally we could use unmodified bullmq without local hacks. Thanks @manast and @thegedge
